### PR TITLE
New OS: Ekinops

### PR DESCRIPTION
**Pull request by h-barnhart**
New device OS: Ekinops Optical Networking.

The Ekinops have a custom SNMP implementation, so please keep that in mind when reviewing the code. Ekinops uses a chassis form factor with various line cards for specific tasks. Ekinops does not pool all OIDs available in the chassis into a single tree to query. Instead you must either 1) set a particular OID to the slot number you want to query via SNMP or 2) Modify the SNMP community string and append the slot number. As neither of these functions are currently available in LibreNMS this initial code pull will only include information available from the management card using the base SNMP community string. Additional functionality will be added over time.

This PR will include collecting ports, line cards, and line card state. It also modifies the port ifDescr and ifAlias to make them easier to read in the UI.

Edit:

This PR also adds trap handlers for log events (Mgnt2TrapNMSEvent) and alarms (Mgnt2TrapNMSAlarm).

Chassis tested by me include a C600CH and C200H.

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by h-barnhart** [comment]
Looks like TravisCI is tripping on my ifDescr changer. I can get rid of the changer if its not acceptable. Without it the interface names are really long:

EKINOPS/C600HC/8/PMROADMC/Express4(PORT_Number 4   )

**Comment by PipoCanaja** [comment]
Hi @h-barnhart 
It seems that your test data is not matching this PR. Could you try to run again save-test-data.php and commit the resulting file ?

**Comment by h-barnhart** [comment]
What I provided includes my code for modifying the interface ifAlias and IfDescr, which doesn't appear to take effect during the data gathering stage. My code for changing those values is in includes/polling/ports/os/ekinops.inc.php. Is there another method I should use or just drop the code for changing the ifDescr and ifAlias?

**Comment by h-barnhart** [comment]
Ok, I had to put my code that modifies ifDescr and ifAlias under the discovery code path as well as the poller. I realize this is code duplication and if the discovery and polling subsystems were OOB it could be avoided. If there is a better way let me know.

Beyond that I think it should checkout now as the snmpsim data will match what is actually collected during discovery.

**Comment by h-barnhart** [comment]
Forgot to add my snmpsim data, only updated the json

File: includes/polling/os/ekinops.inc.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?php
```
Comment (Jellyfrog) [inline includes/polling/os/ekinops.inc.php:1]: This file should be converted to the new-ish format, check here for more info, https://docs.librenms.org/Developing/os/Initial-Detection/#os-discovery

Comment (Jellyfrog) [inline includes/polling/os/ekinops.inc.php:1]: This file should be converted to the new-ish format, check here for more info, https://docs.librenms.org/Developing/os/Initial-Detection/#os-discovery

**Review by Jellyfrog — CHANGES_REQUESTED** [review]

**Comment by h-barnhart** [comment]
Figured out the issue with the SNMP test data. First of all, some of the bad-if settings were preventing all my interfaces from being discovered, so they were in the snmpsim but not the json. Second I was running the save-test-data script on my test server and uploading the JSON to my dev server. I tried uploading just the snmpsim and then running save-test-data from the dev server and now unit tests are passing on my dev install. I'm going to let the checks run on GitHub and go from there.

Just an FYI, on my dev install some other OS's were failing unit tests, nothing I can do about that.

@Jellyfrog If travis checks out I'll go back and re-do the code you pointed out.

File: includes/polling/os/ekinops.inc.php (hunk 1 above)
Comment (murrant) [reply includes/polling/os/ekinops.inc.php:1]: Should be able to do it all in yaml now. (use regex to parse mgnt2RinvHwPlatform)

File: includes/discovery/sensors/pre-cache/ekinops.inc.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * Traps when Adva objects are created. This includes Remote User Login object,
+ * Flow Creation object, and LAG Creation object.
+ *
+ * @package    LibreNMS
+ * @link       http://librenms.org
+ * @copyright  2020 KanREN, Inc
+ * @author     Heath Barnhart <hbarnhart@kanren.net>
+ */
+
+$pre_cache['ekishelf'] = snmpwalk_cache_multi_oid($device, 'mgnt2GigmBoardTable', [], 'EKINOPS-MGNT2-MIB', null, '-OQUbs');
```
Comment (murrant) [inline includes/discovery/sensors/pre-cache/ekinops.inc.php:32]: Why do you have  prefetch?  I can't see where it is used.

Comment (h-barnhart) [reply includes/discovery/sensors/pre-cache/ekinops.inc.php:32]: Its used for sensor discovery

Comment (h-barnhart) [reply includes/discovery/sensors/pre-cache/ekinops.inc.php:32]: The line below it doesn't need to be there, though.

**Comment by h-barnhart** [comment]
The fail in Travis is for a different OS.

**Comment by PipoCanaja** [comment]
@h-barnhart Could you rebase your PR, a big change was merged earlier today.

File: includes/discovery/sensors/pre-cache/ekinops.inc.php (hunk 2 above)
Comment (murrant) [reply includes/discovery/sensors/pre-cache/ekinops.inc.php:32]: this prefetch doesn't work there and all it does is fetch the data twice and only use it once.  (also the sensor code doesn't need prefetch it is extremely smart about how it fetches the data)

Comment (h-barnhart) [reply includes/discovery/sensors/pre-cache/ekinops.inc.php:32]: I'll pull it from m

**Comment by h-barnhart** [comment]
Do I need to make the changes from styleci?

File: tests/data/ekinops.json [hunk 3]
```diff
@@ -0,0 +1,10572 @@
+{
+    "os": {
+        "discovery": {
+            "devices": [
+                {
+                    "sysName": "<private>",
+                    "sysObjectID": ".1.3.6.1.4.1.20044",
+                    "sysDescr": "Ekinops 360 Platform, C600HC, Release 6.3.900",
+                    "sysContact": null,
+                    "version": " Release 6.3.900",
```
Comment (Jellyfrog) [inline tests/data/ekinops.json:10]: Maybe want to use trim() to get rid of the spaces

Comment (h-barnhart) [reply tests/data/ekinops.json]: fixed

**Comment by Jellyfrog** [comment]
Do we really need all these MIBs? I'm only seeing 2 used

**Comment by h-barnhart** [comment]
I can pull the mibs that aren't needed right now, its more than two when you look at dependencies and the sensors.

**Review by Jellyfrog — APPROVED** [review]

**Comment by Jellyfrog** [comment]
Nice work!

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-68-release-changelog-september-2020/13525/1

**Pull request merged**